crates/ruff_server/src/server/api/requests/execute_command.rs

```diff
@@ -0,0 +1,213 @@
+use std::{collections::HashMap, str::FromStr};
+
+use crate::server::api::LSPResult;
+use crate::server::client;
+use crate::server::schedule::Task;
+use crate::session::Session;
+use crate::DIAGNOSTIC_NAME;
+use crate::{edit::DocumentVersion, server};
+use lsp_server::ErrorCode;
+use lsp_types::{self as types, request as req};
+use serde::Deserialize;
+use types::TextDocumentEdit;
+
+#[derive(Debug)]
+enum Command {
+    Format,
+    FixAll,
+    OrganizeImports,
+}
+
+pub(crate) struct ExecuteCommand;
+
+#[derive(Deserialize)]
+struct TextDocumentArgument {
+    uri: types::Url,
+    version: DocumentVersion,
+}
+
+#[derive(Debug)]
+enum CommandEditTracker {
+    DocumentChanges(Vec<types::TextDocumentEdit>),
+    Changes(HashMap<types::Url, Vec<types::TextEdit>>),
+}
+
+impl super::RequestHandler for ExecuteCommand {
+    type RequestType = req::ExecuteCommand;
+}
+
+impl super::SyncRequestHandler for ExecuteCommand {
+    fn run(
+        session: &mut Session,
+        _notifier: client::Notifier,
+        requester: &mut client::Requester,
+        params: types::ExecuteCommandParams,
+    ) -> server::Result<Option<serde_json::Value>> {
+        let command =
+            Command::from_str(&params.command).with_failure_code(ErrorCode::InvalidParams)?;
+
+        // check if we can apply a workspace edit
+        if !session.resolved_client_capabilities().apply_edit {
+            return Err(anyhow::anyhow!("Cannot execute the '{}' command: the client does not support `workspace/applyEdit`", command.label())).with_failure_code(ErrorCode::InternalError);
+        }
+
+        let mut edit_tracker =
+            CommandEditTracker::new(session.resolved_client_capabilities().document_changes);
+        for arg in params.arguments {
```

Comment:
So is it guaranteed that documents aren't repeated here?

Reply:
We enforce it indirectly by preventing re-setting edits on a document, but we should probably enforce it here too.